File: 31-next-permutation/31-next-permutation.cs
public class Solution {
    public void NextPermutation(int[] nums) {
            int len = nums.Length;
            if (len < 2)
                return;

            int j = len - 2;
            while (j >= 0 && nums[j] >= nums[j+1]) 
                j--;
        
            if (j < 0) {
                Reverse(nums, 0, len-1);
                return;
            }
        
            int l = len - 1;
            while (nums[j] >= nums[l]) 
                l--;
        
            Swap(nums, j, l);
            Reverse(nums, j+1, len-1); 
        }

        private void Reverse(int[] arr, int l, int r) {
            while (l < r) {
                Swap(arr, l, r);
                l++;
                r--;
            }        
        }

        private void Swap(int[] arr, int i, int j) {
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
}

File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int ind=-1;
        int val = 0;
        for(int i=n-2;i>=0;i--){
            if(nums[i]<nums[i+1]){
                ind=i;
                val=nums[i];
                break;
            }
        }
        if(ind==-1){
            reverse(begin(nums),end(nums));
            return;
        }
        int j;
        for(j=n-1;j>ind;j--){
            if(nums[j]>nums[ind]){
                break;
            }
        }
        swap(nums[j],nums[ind]);
        reverse(begin(nums)+ind+1,end(nums));
    }
};